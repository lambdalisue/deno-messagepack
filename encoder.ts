import { Packr } from "msgpackr";

const packr = new Packr({
  useRecords: false,
  sequential: true,
});

/** Encode `value` into a single MessagePack-encoded object and returns `Uint8Array`. */
export function encode<T extends unknown>(value: T): Uint8Array {
  return packr.pack(value);
}

/** Transform a stream into a stream where each chunk is encoded into a single MessagePack-encoded object.
 *
 * ```ts
 * import { EncodeStream } from "./encoder.ts";
 *
 * const stream = new ReadableStream({
 *   start(controller) {
 *     controller.enqueue({ foo: "bar" });
 *     controller.enqueue({ qux: "quux" });
 *     controller.close();
 *   },
 * });
 * const messages = stream.pipeThrough(new EncodeStream());
 * ```
 */
export class EncodeStream<T extends unknown>
  extends TransformStream<T, Uint8Array> {
  constructor() {
    super({
      transform: (chunk, controller) => {
        controller.enqueue(encode<T>(chunk));
      },
    });
  }
}
