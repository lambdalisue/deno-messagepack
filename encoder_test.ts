import { encode, EncodeStream } from "./encoder.ts";
import { assertSnapshot } from "@std/testing/snapshot";

const testcases = {
  string: "foo",
  number: 100,
  array: ["a", 0, "b", 1, "c", 2],
  object: { foo: "bar", baz: "qux" },
};

Deno.test("encode", async (t) => {
  for (const [key, value] of Object.entries(testcases)) {
    await t.step(key, async () => {
      const result = encode(value);
      await assertSnapshot(t, result);
    });
  }
});

Deno.test("EncodeStream", async (t) => {
  const stream = new ReadableStream({
    start(controller) {
      for (const testcase of Object.values(testcases)) {
        controller.enqueue(testcase);
      }
      controller.close();
    },
  });
  const chunks = stream.pipeThrough(new EncodeStream());
  for await (const chunk of chunks) {
    await assertSnapshot(t, chunk);
  }
});
