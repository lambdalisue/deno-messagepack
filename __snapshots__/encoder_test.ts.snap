export const snapshot = {};

snapshot[`EncodeStream 1`] = `<Buffer a3 66 6f 6f>`;

snapshot[`EncodeStream 2`] = `<Buffer 64>`;

snapshot[`EncodeStream 3`] = `<Buffer 96 a1 61 00 a1 62 01 a1 63 02>`;

snapshot[`EncodeStream 4`] = `<Buffer de 00 02 a3 66 6f 6f a3 62 61 72 a3 62 61 7a a3 71 75 78>`;

snapshot[`encode 1`] = `<Buffer a3 66 6f 6f>`;

snapshot[`encode 2`] = `<Buffer 64>`;

snapshot[`encode 3`] = `<Buffer 96 a1 61 00 a1 62 01 a1 63 02>`;

snapshot[`encode 4`] = `<Buffer de 00 02 a3 66 6f 6f a3 62 61 72 a3 62 61 7a a3 71 75 78>`;
